### PR TITLE
Sized String Support = Optimizations

### PR DESCRIPTION
**Pull request by manuel-fischer**
Added sized string support, such that it can be used for example with C++ `std::string_view`. Introducing an internal `StringView` struct, that contains a pointer, the size of the string and the encoding.

This allows to reduce string allocations, which were needed to convert to UTF8 and for splitting the strings in wrapped rendering. Instead I have solved both problems by almost always using sized strings and by moving the encoding specific string operations deeper into the call chain.

Also the way how layouting with the position buffer works is changed. The arrays are extracted from the font struct into an own position array, the struct contains a dummy element for ABI stability. The new array-struct is passed as an explicit parameter, which can be left `NULL`, to avoid allocation in `TTF_Size_Internal`.

## New functions
Each function related to rendering text has added an additional parameter `size_t text_size` after the `text` parameter.
- `TTF_Size*_SZ`
- `TTF_Measure*_SZ`
- `TTF_Render*_Solid_SZ`
- `TTF_Render*_Shaded_SZ`
- `TTF_Render*_Blended_SZ`
- `TTF_Render*_Solid_Wrapped_SZ`
- `TTF_Render*_Shaded_Wrapped_SZ`
- `TTF_Render*_Blended_Wrapped_SZ`

`*` represents all of `TEXT`, `UTF8`, `UNICODE`

## For the documentation of the new functions
For UTF8 and TEXT `text_size` denotes the number of bytes, for UNICODE it denotes the number of words (number of bytes / 2)

`NULL` can be passed as a paramter, as long as the size is also `0`.
The range `[text, text+text_size)` should be a valid string and thus a valid memory range.

## Showfont
Added new command line parameter `[-wrap width]` which forces to use the `*_Wrapped` functions with `width` as the `wrapLength` parameter.

## Problem
One problem, that might need to be discussed is the combinatoric explosion of rendering functions for the different formats.

### CODE REVIEW

**Comment by sezero** [comment]
The patch does not apply: you generated against an older git revision. It seems like you miss this commit: https://github.com/libsdl-org/SDL_ttf/commit/8368ea92c8bb86d65dd28ba10eeeeecefbd7d214

File: SDL_ttf.c [hunk 1]
```diff
@@ -3329,82 +3516,167 @@ SDL_Surface* TTF_RenderGlyph32_Shaded(TTF_Font *font, Uint32 ch, SDL_Color fg, S
...
+            }
+            new_lines = (TTF_StringView *)SDL_realloc(*p_lines, maxNumLines * sizeof (*p_lines));
+            if (new_lines == NULL) {
+                SDL_OutOfMemory();
+                return SDL_FALSE;
+            }
+            *p_lines = new_lines;
+        }
+
+        TTF_StringView *line = &(*p_lines)[*p_numLines];
```
Comment (sezero) [inline SDL_ttf.c:3598]: This will hit `-Werror=declaration-after-statement`

**Comment by manuel-fischer** [comment]
> The patch does not apply: you generated against an older git revision. It seems like you miss this commit: [8368ea9](https://github.com/libsdl-org/SDL_ttf/commit/8368ea92c8bb86d65dd28ba10eeeeecefbd7d214)

It was a bug, that I have found while rewriting the code, it can be discarded, because I have rewritten that part in the code.

File: showfont.c [hunk 2]
```diff
@@ -266,36 +277,70 @@ int main(int argc, char *argv[])
...
     /* Render and center the message */
-    if (argc > 2) {
-        message = argv[2];
+    if (argc >= 2) {
+        if(wrap) // concatenate message
+        {
+            size_t start = 0;
+            size_t rest_bytes = sizeof(string);
+            
+            for(int i = 2; i < argc && rest_bytes > 0; ++i)
```
Comment (sezero) [inline showfont.c:286]: This will fail build in C90 mode.

**Comment by sezero** [comment]
>  it can be discarded, because I have rewritten that part in the code.

You should rebase your patchset, instead.

File: SDL_ttf.c [hunk 3]
```diff
@@ -2985,13 +3101,10 @@ static int TTF_Size_Internal(TTF_Font *font,
         int y_offset  = hb_glyph_position[g].y_offset;
 #else
     /* Load each character and sum it's bounding box */
-    textlen = SDL_strlen(text);
```
Comment (sezero) [inline SDL_ttf.c:2988]: The local `size_t textlen;` should be removed from the top of the procedure, as well.

**Comment by 1bsyl** [comment]
I also want to look at this (end of next week probably).
Quickly:
- there are still things commented out you could remove (About pa allocation).
- do we really need a posbuf array.(is this reallocated more than once, since we use a new one each time). But I may need to read more precisely

**Comment by manuel-fischer** [comment]
>     * do we really need a posbuf array.(is this reallocated more than once, since we use a new one each time). But I may need to read more precisely

Multiple Reallocations do happen for the posbuf, while previously the posbuf is permanently saved in the font and not resized, if it was large enough. But this could be improved by either allocating directly by the amount of glyphs in the string or by rewriting it to not use allocated arrays at and iterate again through the string characters. Note that for the latter, the glyph-lookup would not change, because it is already looked up again, when iterating through the posbuf-array.

**Comment by 1bsyl** [comment]
ok, I've looked and tried (but not fully)
Just to make sure I saw no regression with some multi lang test I have.
I also run a fuzzy test, and it didn't crash.

Maybe not a problem, but there is lot of modification in the wrap version. and there were corner cases and special behavior (with wrap-length == 0 for instance). hope we have no regression.

I see now it decodes  unicode and text, instead of converting to utf8, which somehow requires more testing in case of bug. So not sure if this is better.
As you said, API is longer with all _SZ variants for string_view.

Measuring the time with some bench, I didn't see a noticeable speed improvement,
so I wonder if this worth adding more complexity in sdl_ttf code
and if string_view API could just be built on top of RenderUTF8 function.

**Comment by slouken** [comment]
SDL_ttf has been updated to remove Latin-1 and UNICODE API surfaces and support strings with length by default.
